src/main/java/org/broadinstitute/consent/http/service/DatasetRegistrationService.java

```
@@ -0,0 +1,426 @@
+
+/**
+ * Specialized service class which specifically handles
+ * the process of registering a new dataset into the system,
+ * i.e. creates new datasets.
+ */
+public class DatasetRegistrationService {
+    private final DatasetDAO datasetDAO;
+    private final DacDAO dacDAO;
+    private final DatasetServiceDAO datasetServiceDAO;
+    private final GCSService gcsService;
+
+    public DatasetRegistrationService(DatasetDAO datasetDAO, DacDAO dacDAO, DatasetServiceDAO datasetServiceDAO, GCSService gcsService) {
+        this.datasetDAO = datasetDAO;
+        this.dacDAO = dacDAO;
+        this.datasetServiceDAO = datasetServiceDAO;
+        this.gcsService = gcsService;
+    }
+
+
+    /**
+     * This method takes an instance of a dataset registration schema and creates datasets from it.
+     * There will be one dataset per ConsentGroup in the dataset.
+     *
+     * @param registration The DatasetRegistrationSchemaV1.yaml
+     * @param user The User creating these datasets
+     * @param files Map of files, where the key is the name of the field
+     * @return List of created Datasets from the provided registration schema
+     */
+    public List<Dataset> createDatasetsFromRegistration(
+            DatasetRegistrationSchemaV1 registration,
+            User user,
+            Map<String, FormDataBodyPart> files) throws IOException, SQLException, IllegalArgumentException {
+
+        if (Objects.isNull(dacDAO.findById(registration.getDataAccessCommitteeId()))) {
+            throw new IllegalArgumentException("Invalid DAC id");
+        }
+
+        Map<String, BlobId> uploadedFileCache = new HashMap<>();
+
+        List<DatasetServiceDAO.DatasetInsert> datasetInserts = new ArrayList<>();
+        for (int consentGroupIdx = 0; consentGroupIdx < registration.getConsentGroups().size(); consentGroupIdx++) {
+            datasetInserts.add(createDatasetInsert(registration, user, files, uploadedFileCache, consentGroupIdx));
+        }
```

I generally lean towards lambdas for this kind of thing, but `createDatasetInsert` throws `IOException` so it doesn't work inside the lambda unless it's re-thrown as a `RuntimeException`. So, if we hit an exception on index 3 out of 5, for example, then 0, 1, and 2 would have inserted files via `uploadFile` which is called inside the loop. Then we'd have `N` files left behind.

I think one solution to this problem would be to split up the file uploads into one block, and then, if they all succeed, proceed with the dataset inserts using the cached files stored in uploadedFileCache. That would allow for:

1. rolling back the file uploads if any fail
2. generating the inserts without throwing an `IOException`
3. rolling back file uploads if the dataset inserts fail

---

I originally did that solution, but I didn't want to upload all files, because it was possible that the user uploaded a file we don't need, and I didn't want to upload anything that we wouldn't use. I also think a lambda here would muddy up the function, since `createDatasetInsert` is a non-trivial function. The function is already doing a lot, I think. I do agree on rolling back files though, I can add some try/catch blocks to delete the files if something goes wrong.

---

My thought on the lambda was not to replace `createDatasetInsert`, but to call that function inside the lambda instead of using placeholder variables and mutating lists - this is where we often find bugs. The blocker for that was the `IOException`, hence my suggestion to get around that with pre-uploads. One (probably very) minor benefit to the separate upload/dataset insert is that you're not blocking the open db handle while file uploads to GCS are happening. Breaking those two things into discrete chunks will ease the pressure on the handle. 

That said, I can live with this approach :-)

---

Can you make the DB bits transactional so that the only rollback would be deleting the orphaned files? (and also you could make `createDatasetInsert` no throw a checked exception so that you could call it in a lambda...Man I with there was a way to handle errors in lambdas more cleanly)

---

> Can you make the DB bits transactional so that the only rollback would be deleting the orphaned files? (and also you could make `createDatasetInsert` no throw a checked exception so that you could call it in a lambda...Man I with there was a way to handle errors in lambdas more cleanly)

The DB bits are transactional, but we have two systems to save to, so combining them is the thing that's problematic to me. We do the following in a single block:
1. Save a dataset with file metadata in postgres (transactional and rollback-able)
2. Save the actual file content in GCS

What I would prefer would be to break them out into two separate rollback-able things so we can fail faster if there are problems in either logical block.

---

To clarify... creating the dataset insert is just creating a record object. So there is no database transaction happening. Basically, what's happening:

1. Upload all files & create dataset insert records at same time (no db changes)
2. Commit all dataset inserts in a rollback-able way at the same time

They are happening sequentially, the only thing I'm missing (if I understand the discourse here) is the ability to rollback files.

---

I added rollback for files. I am gonna add a few tests, and I'll also add some tests for file rollback.